### PR TITLE
Spia

### PR DESCRIPTION
**Pull request by ddomingof**
Add SPIA module for exporting BELGraphs to SPIA software

### CODE REVIEW

**Comment by ddomingof** [comment]
@cthoyt how to merge it?

File: src/pybel_tools/analysis/spia.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    """Return set of HGNC names from Proteins/Rnas/Genes/miRNA, nodes that can be used by SPIA.
+
+    :param graph: BELGraph
+    :return: node names
+    """
+    # TODO: Using HGNC Symbols for now
+    return {
+        node.name
+        for node in graph.nodes()
+        if issubclass(type(node), BaseAbundance) and node.namespace == 'HGNC'
```
Comment (cthoyt) [inline src/pybel_tools/analysis/spia.py]: This is a strange way of doing this... Why not`isinstance(node, BaseAbundance)`?

File: src/pybel_tools/analysis/spia.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+import pandas as pd
+
+from pybel import BELGraph
+from pybel.constants import ASSOCIATION, INCREASES, DECREASES, DIRECTLY_DECREASES, DIRECTLY_INCREASES
+from pybel.dsl.node_classes import BaseAbundance, CentralDogma, Gene, ListAbundance, Rna, ProteinModification
+from pybel_tools.constants import KEGG_RELATIONS
+
+
+def get_matrix_index(graph: BELGraph) -> set:
```
Comment (cthoyt) [inline src/pybel_tools/analysis/spia.py]: Set of what? Use `typing.Set`

File: src/pybel_tools/analysis/spia.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    """Return set of HGNC names from Proteins/Rnas/Genes/miRNA, nodes that can be used by SPIA.
+
+    :param graph: BELGraph
```
Comment (cthoyt) [inline src/pybel_tools/analysis/spia.py]: Not really helpful comments on parameters... just skip them

File: src/pybel_tools/analysis/spia.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    """
+    # TODO: Using HGNC Symbols for now
+    return {
+        node.name
+        for node in graph.nodes()
+        if issubclass(type(node), BaseAbundance) and node.namespace == 'HGNC'
+    }
+
+
+def build_matrices(nodes: set) -> Dict[str, pd.DataFrame]:
```
Comment (cthoyt) [inline src/pybel_tools/analysis/spia.py]: fix type annotation for set

File: src/pybel_tools/analysis/spia.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    :param nodes: nodes to be included in the matrix
+    :return: Dictionary of adjency matrix for each relationship
+    """
+    return {
+        relation: pd.DataFrame(0, index=nodes, columns=nodes)
+        for relation in KEGG_RELATIONS
+    }
+
+
+def update_matrix(matrix_dict, sub, obj, data):
```
Comment (cthoyt) [inline src/pybel_tools/analysis/spia.py:39]: Docstrings?

File: src/pybel_tools/analysis/spia.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        relation = data['relation']
+
+        if relation in {DIRECTLY_INCREASES, INCREASES}:
+
+            # If it has pmod check which one and add it to the corresponding matrix
+            if obj.variants and any(isinstance(variant, ProteinModification) for variant in obj.variants):
+
+                for variant in obj.variants:
+
+                    if variant['identifier']['name'] == "Ub":
```
Comment (cthoyt) [inline src/pybel_tools/analysis/spia.py]: use pybel constants

File: src/pybel_tools/analysis/spia.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+                        matrix_dict["activation_ubiquination"][subject_name][object_name] = 1
+
+                    elif variant['identifier']['name'] == "Ph":
```
Comment (cthoyt) [inline src/pybel_tools/analysis/spia.py]: make generic functions for testing phosphorylation, especially if there are multiple valid possibilities (what about equivalent GO term?)

File: src/pybel_tools/analysis/spia.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+                        matrix_dict["activation_phosphorylation"][subject_name][object_name] = 1
+
+            # Normal increase, add activation
+            else:
+
+                if isinstance(obj, Gene) or isinstance(obj, Rna):
```
Comment (cthoyt) [inline src/pybel_tools/analysis/spia.py]: `isinstance(obj, (Gene, Rna))`

File: src/pybel_tools/analysis/spia.py (hunk 1 above)
Comment (ddomingof) [reply src/pybel_tools/analysis/spia.py]: It is the same but ok...

File: src/pybel_tools/analysis/spia.py (hunk 7 above)
Comment (ddomingof) [reply src/pybel_tools/analysis/spia.py]: Can you list all possibilities? Not sure whether I am aware of all the ones out there...

**Comment by ddomingof** [comment]
Ready to be merged @cthoyt

File: src/pybel_tools/analysis/spia.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+            # Normal decrease, check which matrix
+            else:
+
+                if isinstance(obj, Gene) or isinstance(obj, Rna):
+                    matrix_dict["repression"][subject_name][object_name] = 1
+
+                else:
+                    matrix_dict["inhibition"][subject_name][object_name] = 1
+
+        if relation == ASSOCIATION:
```
Comment (cthoyt) [inline src/pybel_tools/analysis/spia.py:97]: that's not what the association relationship is for...

File: src/pybel_tools/analysis/spia.py (hunk 7 above)
Comment (scolby33) [reply src/pybel_tools/analysis/spia.py]: You should use an `Enum` for all the possiblities @cthoyt

**Pull request merged**